presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -348,14 +395,38 @@ public Type getType()
 
         public void processPage(GroupByIdBlock groupIds, Page page)
         {
-            if (step.isInputRaw()) {
+            if (!this.orderByChannels.isEmpty()) {
+                checkArgument(step == Step.SINGLE && pageIndex.isPresent(), "Order by over distributed aggregation is not supported");
+                pageIndex.ifPresent(value -> value.addPage(page));
+            }
+            else if (step.isInputRaw()) {
                 aggregation.addInput(groupIds, page);
             }
             else {
                 aggregation.addIntermediate(groupIds, page.getBlock(intermediateChannel));
             }
         }
 
+        public void processOrderBy()
```

Comment:
Please extract changes done in this class, to some separate "sorting" class. In the `InMemoryHashAggregationBuilder` constructor you would create:
```
this.mySortingClass = new MySortingClass(orderByChannels, ordering, ...)
```

in `processPage` you would call:
```
this.mySortinClass.addPage(page)
```

and `MySortingClass` would have `Iterator<Page> getSortedPages()`.